lib/seahorn/BvOpSem2TrackingRawMemMgr.hh

```diff
@@ -0,0 +1,443 @@
+    assert(llvm::Log2_64(val) + 1 <= g_MetadataBitWidth);
+    return m_metadata.MemSet(ptr, m_ctx.alu().si(val, g_MetadataBitWidth), len,
+                             memIn.getMetadata(), align);
+  }
+
+  // This function reports a warning if there is a possibily for the
+  // passed value to not be equal to 1.
+  // This implies we tried to load memory that was never stored into.
+  void CheckDefBeforeUse(Expr val) {
+    // assert(toCheck == 1)
+    Expr toCheck = mk<NEQ>(val, m_ctx.alu().si(1, g_MetadataBitWidth));
+    // We need to reset the solver everytime since two checks will
+    // need to remove expressions from the solver.
+    m_ctx.resetSolver();
+
+    for (auto e : m_ctx.side()) {
+      m_ctx.addToSolver(e);
+    }
+    m_ctx.addToSolver(m_ctx.getPathCond());
+    m_ctx.addToSolver(toCheck);
+    auto result = m_ctx.solve();
+    if (result) {
+      const Instruction &I = m_ctx.getCurrentInst();
+      auto dloc = I.getDebugLoc();
+      if (dloc) {
+        WARN << "Found Use before Def!"
+             << "[" << dloc->getFilename() << ":" << dloc->getLine() << "]";
+      } else {
+        WARN << "Found Use before Def!"
+             << "[" << I << "]";
+      }
+    }
+  }
+
+  Expr loadIntFromMem(PtrTy ptr, MemValTy mem, unsigned int byteSz,
+                      uint64_t align) {
+    Expr res = m_main.loadIntFromMem(ptr, mem.getRaw(), byteSz, align);
+    Expr metadata = m_metadata.loadIntFromMem(ptr, mem.getMetadata(),
+                                              g_MetadataByteWidth, align);
+    CheckDefBeforeUse(metadata);
+    return res;
+  }
+
+  PtrTy loadPtrFromMem(PtrTy ptr, MemValTy mem, unsigned int byteSz,
+                       uint64_t align) {
+    PtrTy rawPtr = m_main.loadPtrFromMem(ptr, mem.getRaw(), byteSz, align);
+    Expr metadata =
+        m_metadata.loadIntFromMem(ptr, mem.getMetadata(), byteSz, align);
+    CheckDefBeforeUse(metadata);
+    return rawPtr;
+  }
+
+  MemValTy storeIntToMem(Expr _val, PtrTy ptr, MemValTy mem,
+                         unsigned int byteSz, uint64_t align) {
+    // We expect _val to be a primitive rather than a container
+    assert(!strct::isStructVal(_val));
+    RawMemValTy rawVal =
+        m_main.storeIntToMem(_val, ptr, mem.getRaw(), byteSz, align);
+    // Number of slots to fill with '1' depends on slot size i.e
+    // g_MetadataByteWidth
+    unsigned int len = byteSz / g_MetadataByteWidth;
+    RawMemValTy metadataVal = memsetMetaData(ptr, len, mem, align, 1U);
+    return MemValTy(rawVal, metadataVal);
+  }
+
+  MemValTy storePtrToMem(PtrTy val, PtrTy ptr, MemValTy mem,
+                         unsigned int byteSz, uint64_t align) {
+    RawMemValTy rawVal =
+        m_main.storePtrToMem(val, ptr, mem.getRaw(), byteSz, align);
+    // Number of slots to fill with '1' depends on slot size i.e
+    // g_MetadataByteWidth
+    unsigned int len = byteSz / g_MetadataByteWidth;
+    RawMemValTy metadataVal = memsetMetaData(ptr, len, mem, align, 1U);
```

Comment:
eventually we want to set only one byte per object. not clear whether it will clash with other clients